File: 2019/5/intcode.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)

	scanner.Scan()
	program := parseProgram(scanner.Text())

	RunProgram(program, ConsoleInputProvider{}, ConsoleOutputSink{})
}

func parseProgram(input string) []int {
	program := []int{}
	if input[len(input)-1] != ',' {
		input = input + ","
	}

	start := 0
	for index, char := range input {
		if char == ',' {
			i, _ := strconv.Atoi(input[start:index])
			program = append(program, i)
			start = index + 1
		}
	}
	return program
}

type InputProvider interface {
	GetInput() int
}

type OutputSink interface {
	OutputValue(value int)
}

type ConsoleInputProvider struct {}

func (_ ConsoleInputProvider) GetInput() int {
	var i int
	fmt.Scan(&i)
	return i
}

type ConsoleOutputSink struct {
}

func (_ ConsoleOutputSink) OutputValue(value int) {
	fmt.Println(value)
}

// ParameterMode represents how to interpret a parameter
type ParameterMode int

const (
	// Position means interpret the value at the memory position
	Position ParameterMode = iota

	// Immediate means to use the value directly
	Immediate
)

// Instruction represents an IntCode instruction
type Instruction interface {
	Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int
}

// Parameter represents a parameter to an instruction
type Parameter struct {
	Value int
	Mode  ParameterMode
}

func (p Parameter) Load(memory []int) int {
	if p.Mode == Immediate {
		return p.Value
	} else if p.Mode == Position {
		return memory[p.Value]
	}

	// Uh, error?
	return 0
}

func (p Parameter) Store(memory []int, value int) {
	if p.Mode != Position {
		// Uh, add error handling?
	}

	memory[p.Value] = value
}

// StopInstruction represents a stop instruction (opcode 99)
type StopInstruction struct {
}

func (si StopInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	return -1
}

type AddInstruction struct {
	Input1 Parameter
	Input2 Parameter
	Output Parameter
}

func (ai AddInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	input1 := ai.Input1.Load(memory)
	input2 := ai.Input2.Load(memory)
	result := input1 + input2
	ai.Output.Store(memory, result)
	return programCounter + 4
}

type MultiplyInstruction struct {
	Input1 Parameter
	Input2 Parameter
	Output Parameter
}

func (mi MultiplyInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	input1 := mi.Input1.Load(memory)
	input2 := mi.Input2.Load(memory)
	result := input1 * input2
	mi.Output.Store(memory, result)
	return programCounter + 4
}

type InputInstruction struct {
	Position Parameter
}

func (ii InputInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	ii.Position.Store(memory, inputProvider.GetInput())
	return programCounter + 2
}

type OutputInstruction struct {
	Position Parameter
}

func (oi OutputInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	outputSink.OutputValue(oi.Position.Load(memory))
	return programCounter + 2
}

type JumpIfTrueInstruction struct {
	ValueToTest Parameter
	Destination Parameter
}

func (jiti JumpIfTrueInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	valueToTest := jiti.ValueToTest.Load(memory)
	destination := jiti.Destination.Load(memory)
	if valueToTest != 0 {
		return destination
	}

	return programCounter + 3
}

type JumpIfFalseInstruction struct {
	ValueToTest Parameter
	Destination Parameter
}

func (jifi JumpIfFalseInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	valueToTest := jifi.ValueToTest.Load(memory)
	destination := jifi.Destination.Load(memory)
	if valueToTest == 0 {
		return destination
	}
	
	return programCounter +3
}

type LessThanInstruction struct {
	Input1 Parameter
	Input2 Parameter
	Output Parameter
}

func (lti LessThanInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	input1 := lti.Input1.Load(memory)
	input2 := lti.Input2.Load(memory)
	if input1 < input2 {
		lti.Output.Store(memory, 1)
	} else {
		lti.Output.Store(memory, 0)
	}
	return programCounter + 4
}

type EqualsInstruction struct {
	Input1 Parameter
	Input2 Parameter
	Output Parameter
}

func (ei EqualsInstruction) Execute(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	input1 := ei.Input1.Load(memory)
	input2 := ei.Input2.Load(memory)
	if input1 == input2 {
		ei.Output.Store(memory, 1)
	} else {
		ei.Output.Store(memory, 0)
	}
	return programCounter + 4
}

// ExecuteNextInstruction extracts the next instruction, and executes it.
// returns a negative number if the program is finished, otherwise the new value of the programCounter
func ExecuteNextInstruction(memory []int, programCounter int, inputProvider InputProvider, outputSink OutputSink) int {
	instruction := parseInstruction(memory, programCounter)

	return instruction.Execute(memory, programCounter, inputProvider, outputSink)
}

func parseInstruction(memory []int, programCounter int) Instruction {
	opcodeAndParameterModes := memory[programCounter]
	opcode := opcodeAndParameterModes % 100
	if opcode == 99 {
		return StopInstruction{}
	}

	parameterModes := opcodeAndParameterModes / 100

	switch opcode {
	case 1:
		return AddInstruction{
			Input1: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Input2: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
			Output: Parameter{Value: memory[programCounter+3], Mode: getMode(parameterModes, 2)},
		}
	case 2:
		return MultiplyInstruction{
			Input1: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Input2: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
			Output: Parameter{Value: memory[programCounter+3], Mode: getMode(parameterModes, 2)},
		}
	case 3:
		return InputInstruction{
			Position: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
		}
	case 4:
		return OutputInstruction{
			Position: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
		}
	case 5:
		return JumpIfTrueInstruction{
			ValueToTest: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Destination: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
		}
	case 6:
		return JumpIfFalseInstruction{
			ValueToTest: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Destination: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
		}
		case 7:
		return LessThanInstruction{
			Input1: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Input2: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
			Output: Parameter{Value: memory[programCounter+3], Mode: getMode(parameterModes, 2)},
		}
	case 8:
		return EqualsInstruction{
			Input1: Parameter{Value: memory[programCounter+1], Mode: getMode(parameterModes, 0)},
			Input2: Parameter{Value: memory[programCounter+2], Mode: getMode(parameterModes, 1)},
			Output: Parameter{Value: memory[programCounter+3], Mode: getMode(parameterModes, 2)},
		}
	}

	return nil
}

func getMode(parameterModes int, position int) ParameterMode {
	mask := int(math.Pow10(position))
	parameterModes = parameterModes / mask
	parameterModes = parameterModes % 10
	if parameterModes == 1 {
		return Immediate
	}

	return Position
}

// RunProgram runs an "IntCode" program
func RunProgram(program []int, inputProvider InputProvider, outputSink OutputSink) {
	programCounter := 0
	for {
		programCounter = ExecuteNextInstruction(program, programCounter, inputProvider, outputSink)
		if programCounter < 0 {
			return
		}
	}
}

// Print prints an array
func Print(array []int) string {
	l := len(array)
	if l == 0 {
		return "{}"
	}

	s := "{"
	for i := 0; i < len(array); i++ {
		s = fmt.Sprintf("%s%d, ", s, array[i])
	}
	s = s[0:len(s)-2] + "}"

	return s
}

File: 2019/9/intcode.py
def main():
    computer = IntCode(InputProvider(), OutputSink())
    param = Parameter(42, ParameterMode.immediate)
    print(param)

if __name__ == "__main__":
    main()

class IntCode:
    def __init__(self, input_provider, output_sink):
        self.program_counter = 0
        self.input_provider = input_provider
        self.output_sink = output_sink
        self.relative_base = 0

    def run_to_completion(self, program):
        memory = Memory(program)
        while True:
            self.program_counter = self.execute_next_instruction(memory)
            if self.program_counter < 0:
                return

    def execute_next_instruction(self, memory):
        instruction = self.parse_instruction(memory)
        return instruction.execute(memory)

    def parse_instruction(self, memory):
        opcode_and_parameter_modes = memory.get_at(self.program_counter)
        opcode = opcode_and_parameter_modes % 100
        if opcode == 99:
            return StopInstruction(self)

        parameter_modes = opcode_and_parameter_modes / 100

        if opcode == 1:
            return AddInstruction(self,
                param1=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                param2=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)),
                output=Parameter(memory.get_at(self.program_counter + 3), self.get_mode(parameter_modes, 3)))
        elif opcode == 2:
            return MultiplyInstruction(self,
                param1=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                param2=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)),
                output=Parameter(memory.get_at(self.program_counter + 3), self.get_mode(parameter_modes, 3)))
        elif opcode == 3:
            return InputInstruction(self,
                position=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)))
        elif opcode == 4:
            return OutputInstruction(self,
                position=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)))
        elif opcode == 5:
            return JumpIfTrueInstruction(self,
                value=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                dest=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)))
        elif opcode == 6:
            return JumpIfFalseInstruction(self,
                value=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                dest=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)))
        elif opcode == 7:
            return LessThanInstruction(self,
                param1=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                param2=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)),
                output=Parameter(memory.get_at(self.program_counter + 3), self.get_mode(parameter_modes, 3)))
        elif opcode == 8:
            return EqualsInstruction(self,
                param1=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)),
                param2=Parameter(memory.get_at(self.program_counter + 2), self.get_mode(parameter_modes, 2)),
                output=Parameter(memory.get_at(self.program_counter + 3), self.get_mode(parameter_modes, 3)))
        elif opcode == 9:
            return AdjustRelativeBaseInstruction(self,
                param=Parameter(memory.get_at(self.program_counter + 1), self.get_mode(parameter_modes, 1)))
        else:
            raise Invalid()

    def get_mode(self, parameter_modes, position):
        position = position - 1
        mask = 10 ** position
        parameter_modes = parameter_modes // mask
        parameter_modes = parameter_modes % 10
        if parameter_modes == 0:
            return ParameterMode.position
        elif parameter_modes == 1:
            return ParameterMode.immediate
        elif parameter_modes == 2:
            return ParameterMode.relative
        else:
            raise Invalid

class ParameterMode:
    immediate = "Immediate"
    position = "Position"
    relative = "relative"

class Invalid(Exception):
    pass

class InputProvider:
    def get_input(self):
        return int(input())

class OutputSink:
    def print_output(self, output):
        print(output)

class Parameter:
    def __init__(self, value, mode):
        self.value = value
        self.mode = mode

    def load(self, memory, relative_base):
        if self.mode == ParameterMode.immediate:
            return self.value
        elif self.mode == ParameterMode.position:
            return memory.get_at(self.value)
        elif self.mode == ParameterMode.relative:
            return memory.get_at(relative_base + self.value)

        raise Invalid()

    def store(self, memory, value, relative_base):
        if self.mode == ParameterMode.position:
            memory.set_at(self.value, value)
        elif self.mode == ParameterMode.relative:
            memory.set_at(relative_base + self.value, value)
        else:
            raise Invalid()

class Memory:
    def __init__(self, program):
        self.program = program
        self.other_values = { }

    def get_at(self, address):
        if address < len(self.program):
            return self.program[address]
        else:
            value = self.other_values.get(address, 0)
            return value

    def set_at(self, address, value):
        if address < len(self.program):
            self.program[address] = value
        else:
            self.other_values[address] = value

class Instruction:
    def __init__(self, computer):
        self.computer = computer

    def execute(self, memory):
        raise NotImplemented

class StopInstruction(Instruction):
    def __init__(self, computer):
        super().__init__(computer)

    def execute(self, memory):
        return -1

class ArithmeticInstruction(Instruction):
    def __init__(self, computer, param1, param2, output):
        super().__init__(computer)
        self.param1 = param1
        self.param2 = param2
        self.output = output

    def execute(self, memory):
        input1 = self.param1.load(memory, self.computer.relative_base)
        input2 = self.param2.load(memory, self.computer.relative_base)
        result = self.operation(input1, input2)
        self.output.store(memory, result, self.computer.relative_base)
        return self.computer.program_counter + 4

    def operation(self, input1, input2):
        raise NotImplementedError

class AddInstruction(ArithmeticInstruction):
    def __init__(self, computer, param1, param2, output):
        super().__init__(computer, param1, param2, output)

    def operation(self, input1, input2):
        return input1 + input2

class MultiplyInstruction(ArithmeticInstruction):
    def __init__(self, computer, param1, param2, output):
        super().__init__(computer, param1, param2, output)

    def operation(self, input1, input2):
        return input1 * input2

class IOInstruction(Instruction):
    def __init__(self, computer, position):
        super().__init__(computer)
        self.position = position

class InputInstruction(IOInstruction):
    def __init__(self, computer, position):
        super().__init__(computer, position)

    def execute(self, memory):
        value = self.computer.input_provider.get_input()
        self.position.store(memory, value, self.computer.relative_base)
        return self.computer.program_counter + 2

class OutputInstruction(IOInstruction):
    def __init__(self, computer, position):
        super().__init__(computer, position)

    def execute(self, memory):
        output = self.position.load(memory, self.computer.relative_base)
        self.computer.output_sink.print_output(output)
        return self.computer.program_counter + 2

class JumpInstruction(Instruction):
    def __init__(self, computer, value, dest):
        super().__init__(computer)
        self.value = value
        self.dest = dest

    def execute(self, memory):
        value = self.value.load(memory, self.computer.relative_base)
        dest = self.dest.load(memory, self.computer.relative_base)
        if self.compare(value):
            return dest
        return self.computer.program_counter + 3

    def compare(self, value):
        raise NotImplementedError

class JumpIfTrueInstruction(JumpInstruction):
    def __init__(self, computer, value, dest):
        super().__init__(computer, value, dest)

    def compare(self, value):
        return value != 0

class JumpIfFalseInstruction(JumpInstruction):
    def __init__(self, computer, value, dest):
        super().__init__(computer, value, dest)

    def compare(self, value):
        return value == 0

class LessThanInstruction(ArithmeticInstruction):
    def __init__(self, computer, param1, param2, output):
        super().__init__(computer, param1, param2, output)

    def operation(self, input1, input2):
        if input1 < input2:
            return 1
        return 0

class EqualsInstruction(ArithmeticInstruction):
    def __init__(self, computer, param1, param2, output):
        super().__init__(computer, param1, param2, output)

    def operation(self, input1, input2):
        if input1 == input2:
            return 1
        return 0

class AdjustRelativeBaseInstruction(Instruction):
    def __init__(self, computer, param):
        super().__init__(computer)
        self.param = param

    def execute(self, memory):
        value = self.param.load(memory, self.computer.relative_base)
        self.computer.relative_base = self.computer.relative_base + value
        return self.computer.program_counter + 2